kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,72 @@
+//! Interface for Digest
+
+use crate::common::leasable_buffer::LeasableBuffer;
+use crate::returncode::ReturnCode;
+
+/// The 'types' of digests, this should define the output size of the digest
+/// operations.
+pub trait DigestType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl DigestType for [u8; 32] {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a, T: DigestType> {
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// On error or success `data` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a digest is computed.
+    /// On error or success `digest` will contain a reference to the original
+    /// digest supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), ReturnCode>, digest: &'static mut T);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a, T: DigestType> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a, T>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
```

Comment:
`original data digest` or `original data`?
How could it contain a (valid) digest before we even call `run` and on top of that in case of error?

Reply:
Whoops, typo here. Fixed.